### PR TITLE
Module 4 - OOP school library: set up associations

### PR DESCRIPTION
**Pull request by ikennarichard**
## Module 4 (Week 3) - OOP school library: set up associations
- Added a `Book`, `Classroom` and `Rental` class
- Implemented the has-many/belongs-to relationship between Classroom and `Student`. 
- Created the many-to-many relationship between Person and Book using the intermediate class Rental.

### CODE REVIEW

File: student.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 require './person'
 
 class Student < Person
+  attr_reader :classroom
+
   def initialize(classroom, age, name = 'Unknown', parent_permission = true)
     super(age, name, parent_permission)
     @classroom = classroom
```
Comment (rloterh) [inline student.rb:8]: - [ ] Over here, we want to establish that _a student belongs to a classroom_. We can achieve that by doing something similar to:

```ruby
class Student < Person
  attr_reader :classroom
  .
  .

  def classroom(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
end
```

Let's read further on the same [here](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships)

File: book.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class Book
+  attr_accessor :title, :author, :rentals
+
+  def initialize(title, author)
+    @title = title
+    @author = author
+    @rentals = []
+  end
```
Comment (rloterh) [inline book.rb:8]: - [ ] As required, let's ensure that a method for adding rental is present. The _book_ itself should be sent as a parameter to create the new _rental_, thereby achieving the _'has-many'_ association. We can achieve this by modifying our code to look like the following example below:

```ruby
class Book
  attr_accessor :title, :author, :rentals
  .
  .
  def add_rental(person, date)
    Rental.new(date, self, person)
  end
end
```

File: rental.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+class Rental
+  attr_accessor :date, :book, :person
+
+  def initialiize(date)
```
Comment (rloterh) [inline rental.rb:4]: - [ ] Lets ensure to add  _Book_ and _Person_  to the constructor parameters so _Book_ and _Person_ are set in it to break the many-to-many relationship, like the code example below: 

```ruby
class Rental
  .
  
  def initialize(date, book, person)
    .
    .

    
  end
end
```

Kindly read further [here](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#many-to-many-relationship)

File: person.rb [hunk 4]
```diff
@@ -6,14 +6,15 @@ def correct_name
...
+  attr_accessor :name, :age, :rentals
   attr_reader :id
 
   def initialize(age, name = 'Unknown', parent_permission = true)
     @id = Random.rand(1...100)
     @name = name
     @age = age
     @parent_permission = parent_permission
+    @rentals = []
   end
```
Comment (rloterh) [inline person.rb:19]: - [ ] Let's ensure a method for adding _rentals_ is present, so as to establish a a _rental_ `belong-to` _person_ relationship, like the code example below:

```ruby
class Person
  attr_accessor :name, :age, :rentals
  .
  .

  def add_rental(book, date)
    Rental.new(date, book, self)
  end

  .
  .
end
```

File: student.rb [hunk 5]
```diff
@@ -1,6 +1,8 @@
 require './person'
```
Comment (rloterh) [inline student.rb:1]: - [ ] [OPTIONAL]
By convention, it's a good practice to use `require_relative` classes in the same source and use `require` for external modules or packages. Kindly ensure to improve upon this aspect.

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Richard, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there!
🏁 🚦 ⛪ 🚜 🏃 🌴 🏫 🌳 🏘️

## Highlights 👍  
- Good job so far with creating associations among objects!
- ReadMe looks good!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

_Kindly consider my inline comments marked **[OPTIONAL]** under this review_

**You can also consider:**
Kindly add instructions in your ReadMe under the _Getting Started_ section, on how to run the project.
Implementing the above change(s) will make your ReadMe/ repository stand out even better! 😉

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: student.rb (hunk 5 above)
Comment (ikennarichard) [reply student.rb:1]: Thank you @rloterh

**Review by zainsadaqat — APPROVED** [review]
# Pull Request Approved ✅ 🎊 🎉 
## Hello @ikennarichard  👋 

### Things to appreciate
- [ ] Good job for implementing the things requested by the previous code reviewer 👍 

### Your project is complete!  There is nothing else to say other than it's time to merge it.

# Congratulations!!! 🎉 🥳 👏 🙌 🎊 ☑️ 



Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**